File: core/core-sync/src/intTestDisabled/java/ch/difty/scipamato/core/sync/jobs/JooqConfigurationTest.java
package ch.difty.scipamato.core.sync.jobs;

import static org.assertj.core.api.Assertions.assertThat;

import java.sql.SQLException;

import com.zaxxer.hikari.HikariDataSource;
import org.jooq.Configuration;
import org.jooq.SQLDialect;
import org.jooq.impl.DataSourceConnectionProvider;
import org.jooq.impl.DefaultExecuteListenerProvider;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.jdbc.datasource.TransactionAwareDataSourceProxy;

@SpringBootTest
class JooqConfigurationTest {

    @Autowired
    @Qualifier("publicConfiguration")
    private Configuration jooqPublicConfig;

    @Autowired
    @Qualifier("coreConfiguration")
    private Configuration jooqCoreConfig;

    @Autowired
    @Qualifier("batchConfiguration")
    private Configuration jooqBatchConfig;

    @Test
    void assertJooqPublicConfigIsProperlyWired() throws SQLException {
        assertConfiguration(jooqPublicConfig);
    }

    @Test
    void assertJooqCoreConfigIsProperlyWired() throws SQLException {
        assertConfiguration(jooqCoreConfig);
    }

    @Test
    void assertJooqBatchConfigIsProperlyWired() throws SQLException {
        assertConfiguration(jooqBatchConfig);
    }

    private void assertConfiguration(Configuration config) throws SQLException {
        config.shouldNotBeNull();

        config.dialect() shouldBeEqualTo SQLDialect.POSTGRES;

        // assert Datasource Connection Provider
        config.connectionProvider().shouldNotBeNull();
        assertThat(config.connectionProvider()).isInstanceOf(DataSourceConnectionProvider.class);
        DataSourceConnectionProvider dscp = (DataSourceConnectionProvider) config.connectionProvider();
        assertThat(dscp.dataSource()).isInstanceOf(TransactionAwareDataSourceProxy.class);
        assertThat(dscp
            .dataSource()
            .isWrapperFor(HikariDataSource.class)).isTrue();

        // assert executeListenerProviders
        config.executeListenerProviders() shouldHaveSize 2;
        DefaultExecuteListenerProvider elp = (DefaultExecuteListenerProvider) config.executeListenerProviders()[0];
        assertThat(elp
            .provide()
            .getClass()
            .getName()).isEqualTo("org.springframework.boot.autoconfigure.jooq.ExceptionTranslatorExecuteListener");

        // assert TransactionProvider
        config.transactionProvider().shouldNotBeNull();
        assertThat(config
            .transactionProvider()
            .getClass()
            .getName()).isEqualTo("org.springframework.boot.autoconfigure.jooq.SpringTransactionProvider");
    }
}

File: public/public-web/src/test/kotlin/ch/difty/scipamato/publ/config/JooqConfigurationTest.kt
package ch.difty.scipamato.publ.config

import com.zaxxer.hikari.HikariDataSource
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeInstanceOf
import org.amshove.kluent.shouldBeTrue
import org.amshove.kluent.shouldHaveSize
import org.amshove.kluent.shouldNotBeNull
import org.jooq.Configuration
import org.jooq.SQLDialect
import org.jooq.impl.DataSourceConnectionProvider
import org.jooq.impl.DefaultExecuteListenerProvider
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.jdbc.datasource.TransactionAwareDataSourceProxy
import java.sql.SQLException

@SpringBootTest
internal class JooqConfigurationTest {

    @Autowired
    private lateinit var jooqConfig: Configuration

    @Test
    @Throws(SQLException::class)
    fun assertJooqConfigIsProperlyWired() {
        jooqConfig.shouldNotBeNull()
        jooqConfig.dialect() shouldBeEqualTo SQLDialect.POSTGRES

        // assert Datasource Connection Provider
        jooqConfig.connectionProvider() shouldBeInstanceOf DataSourceConnectionProvider::class

        val dscp = jooqConfig.connectionProvider() as DataSourceConnectionProvider
        dscp.dataSource() shouldBeInstanceOf TransactionAwareDataSourceProxy::class
        dscp.dataSource().isWrapperFor(HikariDataSource::class.java).shouldBeTrue()

        // assert executeListenerProviders
        jooqConfig.executeListenerProviders() shouldHaveSize 1
        val elp = jooqConfig.executeListenerProviders()[0] as DefaultExecuteListenerProvider
        elp.provide().javaClass.name shouldBeEqualTo "org.springframework.boot.autoconfigure.jooq.JooqExceptionTranslator"

        // assert TransactionProvider
        jooqConfig.transactionProvider().javaClass.name shouldBeEqualTo
            "org.springframework.boot.autoconfigure.jooq.SpringTransactionProvider"
    }
}
